File: java/revenj-core/src/main/java/org/revenj/Revenj.java
package org.revenj;

import org.postgresql.ds.PGPoolingDataSource;
import org.revenj.database.postgres.converters.JsonConverter;
import org.revenj.extensibility.Container;
import org.revenj.extensibility.SystemState;
import org.revenj.serialization.json.DslJsonSerialization;
import org.revenj.patterns.*;
import org.revenj.database.postgres.jinq.JinqMetaModel;
import org.revenj.security.PermissionManager;
import org.revenj.extensibility.PluginLoader;
import org.revenj.extensibility.SystemAspect;
import org.revenj.serialization.Serialization;
import org.revenj.serialization.xml.XmlJaxbSerialization;
import org.w3c.dom.Element;

import javax.sql.DataSource;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.Type;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.sql.Connection;
import java.util.*;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Function;

public abstract class Revenj {
	public static Container setup() throws IOException {
		Properties properties = new Properties();
		File revProps = new File("revenj.properties");
		if (revProps.exists() && revProps.isFile()) {
			properties.load(new FileReader(revProps));
		} else {
			String location = System.getProperty("revenj.properties");
			if (location != null) {
				revProps = new File(location);
				if (revProps.exists() && revProps.isFile()) {
					properties.load(new FileReader(revProps));
				} else {
					throw new IOException("Unable to find revenj.properties in alternative location. Searching in: " + revProps.getAbsolutePath());
				}
			} else {
				throw new IOException("Unable to find revenj.properties. Searching in: " + revProps.getAbsolutePath());
			}
		}
		return setup(properties);
	}

	public static Container setup(Properties properties) throws IOException {
		String plugins = properties.getProperty("revenj.pluginsPath");
		File pluginsPath = null;
		if (plugins != null) {
			File pp = new File(plugins);
			pluginsPath = pp.isDirectory() ? pp : null;
		}
		return setup(
				dataSource(properties),
				properties,
				Optional.ofNullable(pluginsPath),
				Optional.ofNullable(Thread.currentThread().getContextClassLoader()));
	}

	public static DataSource dataSource(Properties properties) throws IOException {
		String jdbcUrl = properties.getProperty("revenj.jdbcUrl");
		if (jdbcUrl == null) {
			throw new IOException("revenj.jdbcUrl is missing from Properties");
		}
		if (!jdbcUrl.startsWith("jdbc:postgresql:")) {
			throw new IOException("Invalid revenj.jdbcUrl provided. Expecting: 'jdbc:postgresql:...'. Found: '" + jdbcUrl + "'.\n" +
					"If you wish to use custom jdbc driver provide custom data source instead of using Postgres builtin data source.");
		}
		org.postgresql.ds.PGPoolingDataSource dataSource = new PGPoolingDataSource();
		dataSource.setUrl(jdbcUrl);
		String user = properties.getProperty("user");
		String revUser = properties.getProperty("revenj.user");
		if (revUser != null && revUser.length() > 0) {
			dataSource.setUser(revUser);
		} else if (user != null && user.length() > 0) {
			dataSource.setUser(user);
		}
		String password = properties.getProperty("password");
		String revPassword = properties.getProperty("revenj.password");
		if (revPassword != null && revPassword.length() > 0) {
			dataSource.setPassword(revPassword);
		} else if (password != null && password.length() > 0) {
			dataSource.setPassword(password);
		}
		return dataSource;
	}

	public static Container setup(
			DataSource dataSource,
			Properties properties,
			Optional<File> pluginsPath,
			Optional<ClassLoader> classLoader) throws IOException {
		ClassLoader loader;
		if (pluginsPath.isPresent()) {
			File[] jars = pluginsPath.get().listFiles(f -> f.getPath().toLowerCase().endsWith(".jar"));
			List<URL> urls = new ArrayList<>(jars.length);
			for (File j : jars) {
				try {
					urls.add(j.toURI().toURL());
				} catch (MalformedURLException ex) {
					throw new IOException(ex);
				}
			}
			loader = classLoader.isPresent()
					? new URLClassLoader(urls.toArray(new URL[urls.size()]), classLoader.get())
					: new URLClassLoader(urls.toArray(new URL[urls.size()]));
		} else if (classLoader.isPresent()) {
			loader = classLoader.get();
		} else {
			loader = Thread.currentThread().getContextClassLoader();
		}
		ServiceLoader<SystemAspect> aspects = ServiceLoader.load(SystemAspect.class, loader);
		return setup(dataSource, properties, Optional.of(loader), aspects.iterator());
	}

	private static class SimpleDomainModel implements DomainModel {

		private String[] namespaces = new String[0];
		private final ClassLoader loader;
		private final ConcurrentMap<String, Class<?>> cache = new ConcurrentHashMap<>();

		SimpleDomainModel(ClassLoader loader) {
			this.loader = loader;
		}

		void setNamespace(String namespaces) {
			String[] parts = namespaces == null ? new String[0] : namespaces.split(",");
			this.namespaces = new String[parts.length];
			for (int i = 0; i < parts.length; i++) {
				String ns = parts[i];
				this.namespaces[i] = ns.length() > 0 ? ns + "." : "";
			}
		}

		@Override
		public Optional<Class<?>> find(String name) {
			if (name == null) {
				return Optional.empty();
			}
			Class<?> found = cache.get(name);
			if (found != null) {
				return Optional.of(found);
			}
			String className = name.indexOf('+') != -1 ? name.replace('+', '$') : name;
			for (String ns : namespaces) {
				try {
					Class<?> manifest = Class.forName(ns + className, true, loader);
					cache.put(name, manifest);
					return Optional.of(manifest);
				} catch (ClassNotFoundException ignore) {
				}
			}
			return Optional.empty();
		}
	}

	public static Container container(boolean resolveUnknown, ClassLoader loader) {
		Container container = new SimpleContainer(resolveUnknown);
		for (SystemAspect aspect : ServiceLoader.load(SystemAspect.class, loader)) {
			try {
				aspect.configure(container);
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}
		return container;
	}

	public static Container setup(
			DataSource dataSource,
			Properties properties,
			Optional<ClassLoader> classLoader,
			Iterator<SystemAspect> aspects) throws IOException {
		RevenjSystemState state = new RevenjSystemState();
		ClassLoader loader = classLoader.orElse(Thread.currentThread().getContextClassLoader());
		SimpleContainer container = new SimpleContainer("true".equals(properties.getProperty("revenj.resolveUnknown")));
		container.registerAs(state, SystemState.class);
		container.registerInstance(properties);
		container.registerInstance(ServiceLocator.class, container, false);
		container.registerInstance(DataSource.class, dataSource, false);
		container.registerInstance(ClassLoader.class, loader, false);
		container.register(GlobalEventStore.class, true);
		container.register(JsonConverter.class, true);
		SimpleDomainModel domainModel = new SimpleDomainModel(loader);
		container.registerInstance(DomainModel.class, domainModel, false);
		container.registerFactory(DataContext.class, LocatorDataContext::asDataContext, false);
		container.registerFactory(UnitOfWork.class, LocatorDataContext::asUnitOfWork, false);
		container.registerFactory(
				new Generic<Function<Connection, DataContext>>(){}.type,
				c -> (Function<Connection, DataContext>) connection -> LocatorDataContext.asDataContext(c, connection),
				false);
		PluginLoader plugins = new ServicesPluginLoader(loader);
		container.registerInstance(PluginLoader.class, plugins, false);
		PostgresDatabaseNotification databaseNotification =
				new PostgresDatabaseNotification(
						dataSource,
						Optional.of(domainModel),
						properties,
						state,
						container);
		container.registerInstance(EagerNotification.class, databaseNotification, false);
		container.registerInstance(DataChangeNotification.class, databaseNotification, true);
		ChangeNotification.registerContainer(container, databaseNotification);
		container.registerGenerics(
				Query.class,
				(c, arr) -> {
					try {
						return c.resolve(SearchableRepository.class, arr[0]).query();
					} catch (ReflectiveOperationException e) {
						throw new RuntimeException(e);
					}
				});
		container.registerFactory(RepositoryBulkReader.class, PostgresBulkReader::create, false);
		container.registerInstance(PermissionManager.class, new RevenjPermissionManager(container), false);
		container.registerClass(new Generic<Serialization<String>>() {
		}.type, DslJsonSerialization.class, false);
		XmlJaxbSerialization xml = new XmlJaxbSerialization(container, Optional.of(plugins));
		container.registerInstance(new Generic<Serialization<Element>>() {
		}.type, xml, false);
		container.registerInstance(xml);
		int total = 0;
		if (aspects != null) {
			JinqMetaModel.configure(container);
			while (aspects.hasNext()) {
				aspects.next().configure(container);
				total++;
			}
		}
		domainModel.setNamespace(properties.getProperty("revenj.namespace"));
		properties.setProperty("revenj.aspectsCount", Integer.toString(total));
		state.started(container);
		return container;
	}

	public static <T extends DomainEvent> void registerEvents(Container container, PluginLoader plugins, Class<T> manifest, Class<T[]> arrayManifest) throws Exception {
		Type gt = Utils.makeGenericType(DomainEventHandler.class, manifest);
		List<Class<DomainEventHandler>> eventHandlers = plugins.find(DomainEventHandler.class, manifest);
		for (Class<DomainEventHandler> h : eventHandlers) {
			container.registerClass(h, h, false);
			container.registerClass(gt, h, false);
		}
		gt = Utils.makeGenericType(DomainEventHandler.class, arrayManifest);
		eventHandlers = plugins.find(DomainEventHandler.class, arrayManifest);
		for (Class<DomainEventHandler> h : eventHandlers) {
			container.registerClass(h, h, false);
			container.registerClass(gt, h, false);
		}
		Type ct = Utils.makeGenericType(Callable.class, manifest);
		gt = Utils.makeGenericType(DomainEventHandler.class, ct);
		eventHandlers = plugins.find(DomainEventHandler.class, ct);
		for (Class<DomainEventHandler> h : eventHandlers) {
			container.registerClass(h, h, false);
			container.registerClass(gt, h, false);
		}
		ct = Utils.makeGenericType(Callable.class, arrayManifest);
		gt = Utils.makeGenericType(DomainEventHandler.class, ct);
		eventHandlers = plugins.find(DomainEventHandler.class, ct);
		for (Class<DomainEventHandler> h : eventHandlers) {
			container.registerClass(h, h, false);
			container.registerClass(gt, h, false);
		}
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/Revenj.scala
package net.revenj

import java.io._
import java.lang.reflect.ParameterizedType
import java.net.{URL, URLClassLoader}
import java.sql.Connection
import java.util.{Properties, ServiceLoader}
import javax.sql.DataSource

import net.revenj.database.postgres.converters.JsonConverter
import net.revenj.extensibility.{Container, PluginLoader, SystemAspect, SystemState}
import net.revenj.patterns._
import net.revenj.serialization.{JacksonSerialization, Serialization}
import org.postgresql.ds.PGPoolingDataSource

import scala.collection.concurrent.TrieMap
import scala.collection.mutable.ArrayBuffer
import scala.concurrent.ExecutionContext
import scala.reflect.runtime.universe._

object Revenj {

  def setup(): Container = {
    val properties = new Properties
    var revProps = new File("revenj.properties")
    if (revProps.exists && revProps.isFile) {
      properties.load(new FileReader(revProps))
    } else {
      val location = System.getProperty("revenj.properties")
      if (location != null) {
        revProps = new File(location)
        if (revProps.exists && revProps.isFile) {
          properties.load(new FileReader(revProps))
        } else {
          throw new IOException(s"Unable to find revenj.properties in alternative location. Searching in: ${revProps.getAbsolutePath}")
        }
      } else {
        throw new IOException(s"Unable to find revenj.properties. Searching in: ${revProps.getAbsolutePath}")
      }
    }
    setup(properties)
  }

  def setup(properties: Properties): Container = {
    val plugins = properties.getProperty("revenj.pluginsPath")
    val pluginsPath = {
      if (plugins != null) {
        val pp = new File(plugins)
        if (pp.isDirectory) Some(pp) else None
      } else None
    }
    setup(dataSource(properties), properties, pluginsPath, Option(Thread.currentThread.getContextClassLoader))
  }

  def dataSource(properties: Properties): DataSource = {
    val jdbcUrl = properties.getProperty("revenj.jdbcUrl")
    if (jdbcUrl == null) {
      throw new IOException("revenj.jdbcUrl is missing from Properties")
    }
    if (!jdbcUrl.startsWith("jdbc:postgresql:")) {
      throw new IOException(s"""Invalid revenj.jdbcUrl provided. Expecting: 'jdbc:postgresql:...'. Found: '$jdbcUrl'.
If you wish to use custom jdbc driver provide custom data source instead of using Postgres builtin data source.""")
    }
    val dataSource = new PGPoolingDataSource
    dataSource.setUrl(jdbcUrl)
    val user = properties.getProperty("user")
    val revUser = properties.getProperty("revenj.user")
    if (revUser != null && revUser.length > 0) {
      dataSource.setUser(revUser)
    } else if (user != null && user.length > 0) {
      dataSource.setUser(user)
    }
    val password = properties.getProperty("password")
    val revPassword = properties.getProperty("revenj.password")
    if (revPassword != null && revPassword.length > 0) {
      dataSource.setPassword(revPassword)
    } else if (password != null && password.length > 0) {
      dataSource.setPassword(password)
    }
    dataSource
  }

  def setup(dataSource: DataSource, properties: Properties, pluginsPath: Option[File] = None, classLoader: Option[ClassLoader] = None, context: Option[ExecutionContext] = None): Container = {
    val loader = {
      if (pluginsPath.isDefined) {
        val jars = pluginsPath.get.listFiles(new FileFilter {
          override def accept(pathname: File): Boolean = {
            pathname.getPath.toLowerCase.endsWith(".jar")
          }
        })
        val urls = if (jars == null) Array.empty[URL] else jars.map(_.toURI.toURL)
        if (classLoader.isDefined) Some(new URLClassLoader(urls, classLoader.get))
        else Some(new URLClassLoader(urls))
      } else if (classLoader.isDefined) {
        classLoader
      } else {
        Option(Thread.currentThread.getContextClassLoader)
      }
    }
    val aspects = ServiceLoader.load(classOf[SystemAspect], loader.orNull).iterator()
    val buf = ArrayBuffer[SystemAspect]()
    while (aspects.hasNext) {
      buf += aspects.next()
    }
    setup(dataSource, properties, loader, context, buf)
  }

  private class SimpleDomainModel(loader: ClassLoader) extends DomainModel {
    private var namespaces = Array.empty[String]
    private val cache = new TrieMap[String, Class[_]]

    def setNamespace(value: String): Unit = {
      val parts = if (value == null) Array.empty[String] else value.split(",").distinct
      namespaces = parts map { ns =>
        if (ns.isEmpty) "" else ns + "."
      }
    }

    def find(name: String): Option[Class[_]] = {
      if (name == null) None
      else cache.get(name) match {
        case res@Some(_) => res
        case _ =>
          val className = if (name.indexOf('+') != -1) name.replace('+', '$') else name
          var found: Option[Class[_]] = None
          namespaces foreach { ns =>
            if (found.isEmpty) {
              try {
                val manifest = Class.forName(ns + className, true, loader)
                cache.put(name, manifest)
                found = Option(manifest)
              } catch {
                case _: Throwable =>
              }
            }
          }
          found
      }
    }
  }

  def container(resolveUnknown: Boolean, loader: ClassLoader): Container = {
    new SimpleContainer(resolveUnknown, loader)
  }

  def setup(
    dataSource: DataSource,
    properties: Properties,
    classLoader: Option[ClassLoader],
    context: Option[ExecutionContext],
    aspects: Seq[SystemAspect]): Container = {

    val state = new RevenjSystemState
    val loader = classLoader.getOrElse(Thread.currentThread.getContextClassLoader)
    val container = new SimpleContainer("true" == properties.getProperty("revenj.resolveUnknown"), loader)
    container.registerInstance(properties)
    container.registerInstance[SystemState](state, handleClose = false)
    container.registerInstance(context.getOrElse(ExecutionContext.global.prepare()))
    container.registerInstance[ServiceLocator](container, handleClose = false)
    container.registerInstance(dataSource, handleClose = false)
    container.registerInstance(loader, handleClose = false)
    container.register[JsonConverter](singleton = true)
    val plugins = new ServicesPluginLoader(loader)
    container.registerInstance[PluginLoader](plugins)
    val domainModel = new SimpleDomainModel(loader)
    container.registerInstance[DomainModel](domainModel, handleClose = false)
    val databaseNotification = new PostgresDatabaseNotification(dataSource, Some(domainModel), properties, state, container)
    container.registerInstance[EagerNotification](databaseNotification, handleClose = false)
    container.registerInstance[DataChangeNotification](databaseNotification, handleClose = true)
    ChangeNotification.registerContainer(container, databaseNotification)
    container.registerAs[JacksonSerialization, JacksonSerialization](singleton = true)
    container.registerAs[Serialization[String], JacksonSerialization](singleton = true)
    container.registerFactory[DataContext](c => LocatorDataContext.asDataContext(c, loader), singleton = false)
    container.registerFactory[UnitOfWork](c => LocatorDataContext.asUnitOfWork(c, loader), singleton = false)
    container.registerFactory[Function1[Connection, DataContext]](c => conn => LocatorDataContext.asDataContext(conn, c, loader), singleton = false)
    aspects foreach { a => a.configure(container) }
    domainModel.setNamespace(properties.getProperty("revenj.namespace"))
    properties.setProperty("revenj.aspectsCount", Integer.toString(aspects.size))
    state.started(container)
    container
  }

  def registerEvents[T <: DomainEvent : TypeTag](container: Container, plugins: PluginLoader, loader: ClassLoader): Unit = {
    val javaClass = Utils.findType(typeOf[T], runtimeMirror(loader)) match {
      case Some(p) =>
        p match {
          case cl: Class[_] => cl
          case _ => throw new IllegalArgumentException(s"Only non-generic types supported. Found: ${typeOf[T]}")
        }
      case p => throw new IllegalArgumentException(s"Unable to detect type: ${typeOf[T]}")
    }
    def processHandlers[X: TypeTag](gt: ParameterizedType, eventHandlers: Seq[Class[DomainEventHandler[X]]]): Unit = {
      for (h <- eventHandlers) {
        container.registerType(h, h, singleton = false)
        container.registerType(gt, h, singleton = false)
      }
    }
    val arrInst = java.lang.reflect.Array.newInstance(javaClass, 0)
    val gt = Utils.makeGenericType(classOf[DomainEventHandler[_]], javaClass)
    val agt = Utils.makeGenericType(classOf[DomainEventHandler[_]], arrInst.getClass)
    val ft = Utils.makeGenericType(classOf[Function0[_]], javaClass)
    val fgt = Utils.makeGenericType(classOf[DomainEventHandler[_]], ft)
    val aft = Utils.makeGenericType(classOf[Function0[_]], arrInst.getClass)
    val afgt = Utils.makeGenericType(classOf[DomainEventHandler[_]], aft)
    val simpleHandlers = plugins.find[DomainEventHandler[T]]
    processHandlers(gt, simpleHandlers)
    val simpleArrayHandlers = plugins.find[DomainEventHandler[Array[T]]]
    processHandlers(agt, simpleArrayHandlers)
    val funcHandlers = plugins.find[DomainEventHandler[Function0[T]]]
    processHandlers(fgt, funcHandlers)
    val arrayFuncHandlers = plugins.find[DomainEventHandler[Function0[Array[T]]]]
    processHandlers(afgt, arrayFuncHandlers)
  }
}
